README.md
# REM

Use REM to send reminders to yourself, or someone else.

Just run the REM daemon on your server then go to the URL with your browser or `POST` to the API instead.

Notifications will be sent to your smartphone if you are using the [Pushover](http://pushover.net) service.

## Installation (Quick and Dirty)

Let's assume you own the domain `cip.li` and want to run the REM daemon on Port 42888 on your RaspberryPi 3

### 1. Download the latest release to your RaspberryPi

```bash
[user@raspberrypi3 ~]$  curl -LO https://github.com/sdaros/rem/releases/download/v0.6.0/rem-Linux-armv7l && chmod +x rem-Linux-armv7l
```

### 2. Customise the config file

Create the file `~/.config/rem/rem.conf` and configure it to suit your needs
```bash
[user@raspberrypi3 ~]$ mkdir -p ~/.config/rem && vim ~/.config/rem/rem.conf
{
	"ApiToken": "n1VrLLmRMPStaX3pA8TPdh2Kl2QS3q", # Needed for the https://pushover.net Notification Service
	"ApiUser": "cf3YtkHfnSQkYb8GTWSZuPrddTPymQ", # Needed for the https://pushover.net Notification Service
	"Domain": "http://cip.li",
	"NotificationApi": "https://api.pushover.net/1/messages.json",
	"Path": "",
	"Port": "42888"
}
```

### 3. Run the binary as a service

```bash
[user@raspberrypi3 ~]$ sudo systemd-run --setenv=HOME=${HOME} --unit=rem ./rem-Linux-armv7l
Running as unit rem.service.
[user@raspberrypi3 ~]$
```

## Installation (Using Digitalocean)

So, now were using a [Digitalocean](https://digitalocean.com) Docker droplet and want the REM daemon to listen on `https://cip.li/rem`.

### 1. Create a Docker Droplet then clone the github repo

Login via ssh and clone the repo to your home directory

```bash
[root@digitalocean ~]$ git clone https://github.com/sdaros/rem ~/rem && cd ~/rem
```

### 2. Customise the config file

Customise `rem.conf`

```bash
[root@digitalocean ~/rem]$ vim rem.conf # Configure it to suit your needs
{
	"ApiToken": "n1VrLLmRMPStaX3pA8TPdh2Kl2QS3q", # Needed for the https://pushover.net Notification Service
	"ApiUser": "cf3YtkHfnSQkYb8GTWSZuPrddTPymQ", # Needed for the https://pushover.net Notification Service
	"Domain": "https://cip.li",
	"NotificationApi": "https://api.pushover.net/1/messages.json",
	"Path": "rem",
	"Port": "443"
}
```

### 3. Configure rem to use [caddy](https://caddyserver.com) as reverse proxy

If you do not already have a reverse proxy setup for your domain name, you can use `docker-compose` to setup caddy to proxy requests for REM

- create a new Caddyfile

```bash
[root@digitalocean ~/rem]$ vim Caddyfile
# A Caddyfile for our example could look like the following:
cip.li {
        proxy /rem rem:443 {
                proxy_header Host {host}
                proxy_header X-Real-IP {remote}
                proxy_header X-Forwarded-Proto {scheme}
        }
        gzip
        tls email_for_lets_encrypt@cip.li
}
```
- download then run `docker-compose` with `-p <YOUR_DOMAIN>`

```bash
[root@digitalocean ~/rem]$ curl -LO $(curl -si https://api.github.com/repos/docker/compose/releases/latest | grep -oP 'http[^"]+'"`uname -s`"-"`uname -m`") > /usr/local/bin/docker-compose && chmod +x /usr/local/bin/docker-compose
[root@digitalocean ~/rem]$ docker-compose -p cipli up -d
```

### 4. If you already have a reverse proxy setup

- run the REM docker image

```bash
[root@digitalocean ~/rem]$ docker run -v ./rem.conf:/app/.config/rem/rem.conf -d -p 42888:42888 --name rem sdaros/rem
```


CHANGELOG.md
# REM Changelog

## 0.6.0 / 2016-09-21

*  Multiple Notifications should now arrive properly on notification device

## 0.5.0 / 2016-05-14

* Detect client timezone on reminder submission

## 0.4.0 / 2016-05-02

* Add Dockerfile

## 0.3.0 / 2016-04-28

* Read configuration parameters from a config file

## 0.2.0 / 2016-04-27

* Provide a form for submitting new reminders

## 0.1.0 / 2016-04-25

* Add reminder using query parameters


rem.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
)

const (
	version = "0.6.0"
)

type App struct {
	Notification
	Domain string
	Path   string
	Port   string
}

func main() {
	app := new(App)
	app.Notification = *&Notification{}
	app.loadConfigurationFile()
	mux := http.NewServeMux()
	mux.Handle("/", CreateReminder(app))
	log.Printf("Serving rem (version: %v) on %v/%v",
		version, ":"+app.Port, app.Path)
	err := http.ListenAndServe(":"+app.Port, mux)
	log.Fatal(err)
}

func (self *App) loadConfigurationFile() {
	homeDir := os.Getenv("HOME")
	configFile, err := os.Open(homeDir + "/.config/rem/rem.conf")
	die("error: unable to find configuration file! %v", err)
	decoder := json.NewDecoder(configFile)
	err = decoder.Decode(&self)
	die("error: unable to parse configuration file: %v", err)
}

func die(format string, err error) {
	if err != nil {
		log.Fatalf(format, err)
	}
}


geoip.go
package main

import (
	"encoding/json"
	"io"
	"log"
	"net"
	"net/http"
	"time"
)

type (
	ip       string
	zone     string
	location string
)

func DefaultTime(req *http.Request) string {
	clientIp := ipFrom(req)
	clientZone := zoneFromGeo(clientIp)
	clientLocation := locationFrom(clientZone)
	clientNow := time.Now().UTC().In(clientLocation)
	return thirtyMinutesAfter(clientNow)
}

func thirtyMinutesAfter(now time.Time) string {
	inThirtyMinutes := now.Add(time.Duration(30) * time.Minute)
	return inThirtyMinutes.Format("15:04")
}

func ipFrom(req *http.Request) ip {
	clientIpWithPort := req.RemoteAddr
	clientIp, _, _ := net.SplitHostPort(clientIpWithPort)
	return ip(clientIp)
}

func zoneFromGeo(clientIp ip) zone {
	resp, err := http.Get("https://freegeoip.net/json/" +
		string(clientIp))
	info("error when geolocating client ip: %v", err)
	defer resp.Body.Close()
	clientZone, err := jsonToZone(resp.Body)
	info("error parsing client ip: %v", err)
	return zone(clientZone)
}

func locationFrom(zn zone) *time.Location {
	loc, err := time.LoadLocation(string(zn))
	info("error processing client time zone: %v", err)
	return loc
}

func jsonToZone(resp io.ReadCloser) (zone, error) {
	type client struct {
		// As specified by freegeoip.net JSON API
		time_zone string
	}
	c := new(client)
	decoder := json.NewDecoder(resp)
	if err := decoder.Decode(&c); err != nil {
		return "", err
	}
	return zone(c.time_zone), nil
}

func info(format string, err error) {
	if err != nil {
		log.Printf(format, err)
	}
}


template.go
package main

type Template struct {
	SuccessMsg string
	InputType  string
	Domain     string
	Path       string
	Port       string
}

func (self *Template) fallbackToFormInputTypeText() {
	// Some versions of android's default browser do
	// not handle <input type="time"> properly.
	self.InputType = "text"
}

var createReminderTemplate = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Send me a reminder</title>
    <meta name="description" content="Add reminders through a simple API ">
    <meta name="viewport" content="width=device-width, initial-scale=1">
  </head>
  <body>
    <h2>Send me a reminder!</h2>
    <p><strong>{{.SuccessMsg}}</strong></p>
    <form id="rem-form" action="{{.Domain}}:{{.Port}}/{{.Path}}" method="POST">
      <div>
        <input id="client-now" type="hidden" name="client-now">
      </div>
      <div>
        <label for="time">Time*: </label>
	<input id="time" type="{{.InputType}}" name="time" required>
      </div>
      <div>
        <label for="message">Message*: </label>
        <input type="text" name="message" required>
      </div>
      <div>
        <label for="date">Date: </label>
        <input id="date" type="date" name="date">
      </div>
      <div><input type="submit" value="Submit"></div>
    </form>
    <script>
      now = new Date();
      document.getElementById("client-now").setAttribute("value", now);
      then = new Date(now.getTime() + 30*60000);
      inThirtyMinutes = addZero(then.getHours()) + ":" + addZero(then.getMinutes());
      dateToday = now.getFullYear() + "-" + addZero(now.getMonth() + 1) + "-" + addZero(now.getDate());
      document.getElementById("time").setAttribute("value", inThirtyMinutes);
      document.getElementById("date").setAttribute("value", dateToday);
      function addZero(i) {
        if (i < 10) {
          i = "0" + i;
        }
        return i;
      }
      document.getElementById("rem-form").reset();
    </script>
  </body>
</html>
`


add_reminder.go
package main

import (
	"bytes"
	"log"
	"net/http"
	"os/exec"
	"strings"
	"time"
)

func AddReminder(app *App) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if isQueryParamsEmpty(r) {
			path := app.Lookup("path").(string)
			http.Redirect(w, r, path+"/create/", http.StatusFound)
		}
		thenDay := r.URL.Query().Get("day")
		thenTime := r.URL.Query().Get("time")
		message := r.URL.Query().Get("message")
		delay, err := timeToSleepFor(thenDay, thenTime)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("Your reminder has been added. Thank you."))
		log.Printf("The reminder '%v' will be sent to you at %v", message, time.Now().Add(delay))
		go func(time.Duration, *App, string) {
			select {
			case <-time.After(delay):
				execute(app, message)
			}
		}(delay, app, message)
	})
}

func execute(app *App, msg string) {
	cmd := exec.Command(app.Lookup("home").(string)+"/bin/rem_script", msg)
	var cmdResult bytes.Buffer
	cmd.Stdout = &cmdResult
	err := cmd.Run()
	if err != nil {
		log.Fatal(err)
	}
	return
}

func timeToSleepFor(thenDay, thenTime string) (time.Duration, error) {
	thenDate, err := thenDate(thenDay, thenTime)
	if err != nil {
		return 0, err
	}
	return thenDate.Sub(time.Now()), nil
}

func thenDate(thenDay, thenTime string) (time.Time, error) {
	cmd := exec.Command("date", "--rfc-2822", "--date", thenDay+" "+thenTime)
	var cmdResult bytes.Buffer
	cmd.Stdout = &cmdResult
	err := cmd.Run()
	if err != nil {
		return time.Now(), err
	}
	then, err := time.Parse(time.RFC1123Z, strings.TrimSuffix(cmdResult.String(), "\n"))
	if err != nil {
		return time.Now(), err
	}
	return then, nil
}

func isQueryParamsEmpty(r *http.Request) bool {
	thenTime := r.URL.Query().Get("time")
	message := r.URL.Query().Get("message")
	if thenTime == "" || message == "" {
		return true
	}
	return false
}


handler.go
package main

import (
	"html/template"
	"log"
	"net/http"
	"strings"
	"time"
)

type Reminder struct {
	Notification
	TemplateData *Template
	ThenDate     string
	ThenTime     string
}

func newReminder(app *App) *Reminder {
	return &Reminder{
		Notification: app.Notification,
		TemplateData: &Template{
			SuccessMsg: "",
			InputType:  "time",
			Domain:     app.Domain,
			Path:       app.Path,
			Port:       app.Port,
		},
	}
}

func CreateReminder(app *App) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		reminder := newReminder(app)
		switch r.Method {
		case "GET":
			if matchesAndroidBrowserUserAgent(r) {
				reminder.TemplateData.fallbackToFormInputTypeText()
			}
			reminder.renderTemplate(w)
			return
		case "POST":
			submit(reminder, w, r)
			return
		}
	})
}

func submit(reminder *Reminder, w http.ResponseWriter, r *http.Request) {
	if err := reminder.validateClientInput(r); err != nil {
		http.Error(w, "Sorry, we were unable to process your Input.",
			http.StatusInternalServerError)
		return
	}
	go send(reminder, r)
	if isAjaxRequest(r) {
		w.WriteHeader(http.StatusOK)
		return
	}
	reminder.TemplateData.SuccessMsg = "Thank you! Your reminder will be sent at " +
		reminder.ThenDate + " " + reminder.ThenTime
	reminder.renderTemplate(w)
	return
}

func send(reminder *Reminder, r *http.Request) {
	delay, err := reminder.calculateNotificationDelay(r)
	if err != nil {
		reminder.notifyAndLogTheError("error: unable to calculate delay:  " + err.Error())
		return
	}
	reminder.logNewReminder()
	select {
	case <-time.After(delay):
		err := reminder.Notification.Notify()
		die("error: unable to use Notification API: %v", err)
		reminder.logSendingReminderNow()
	}
	return
}

func (self *Reminder) calculateNotificationDelay(r *http.Request) (time.Duration, error) {
	clientNow, err := time.Parse("Mon Jan _2 2006 15:04:05 GMT-0700 (MST)",
		r.FormValue("client-now"))
	if err != nil {
		return 0, err
	}
	then, err := time.ParseInLocation("2006-01-02 15:04",
		self.ThenDate+" "+self.ThenTime, clientNow.Location())
	if err != nil {
		return 0, err
	}
	now := time.Now().In(clientNow.Location())
	delay := then.Sub(now)
	return delay, nil
}

func (self *Reminder) renderTemplate(w http.ResponseWriter) {
	tmpl, err := template.New("createReminder").Parse(createReminderTemplate)
	die("error: unable to render template: %v", err)
	w.WriteHeader(http.StatusOK)
	err = tmpl.ExecuteTemplate(w, "createReminder", self.TemplateData)
	die("error: unable to execute template: %v", err)
	return
}

func (self *Reminder) validateClientInput(r *http.Request) error {
	self.ThenDate = r.FormValue("date")
	self.ThenTime = r.FormValue("time")
	self.Message = r.FormValue("message")
	_, err := time.Parse("2006-01-02 15:04", self.ThenDate+" "+self.ThenTime)
	if err != nil {
		return err
	}
	return nil
}

func (self *Reminder) notifyAndLogTheError(err string) {
	log.Printf("%v", err)
	self.Notification.Message = err
	notifyErr := self.Notification.Notify()
	die("%v", notifyErr)
}

func (self *Reminder) logNewReminder() {
	log.Printf("The reminder '%v' will be sent at %v",
		self.Message, self.ThenDate+" "+self.ThenTime)
	return
}

func (self *Reminder) logSendingReminderNow() {
	log.Printf("The reminder '%v' has been sent to Notification API at '%v'",
		self.Message, time.Now())
	return
}

func isAjaxRequest(r *http.Request) bool {
	if r.Header.Get("X-Requested-With") == "XMLHTTPRequest" {
		return true
	}
	return false
}

func matchesAndroidBrowserUserAgent(r *http.Request) bool {
	if strings.Contains(r.Header.Get("User-Agent"), "Android") {
		return true
	}
	return false
}
